discojs/src/models/gpt/layers.ts

```diff
@@ -210,81 +236,197 @@ class GELU extends tf.layers.Layer {
+ * I.e. LMEmbedding is used to translate token indices into token embeddings
+ * as well as to project embeddings back into token indices
+ * The GPT2 model uses the same embedding matrix for both the token embeddings and the language modeling head
+ * Because Tensorflow.js doesn't offer an easy weight sharing mechanism, we need to define a custom layer
+ * that can be used for both the token embeddings and the language modeling head.
+ * In the GPT2 model definition, this layers corresponds to wte and lm_head (which reuses wte)
+ */
+class LMEmbedding extends tf.layers.Layer {
+  static readonly className = 'LMEmbedding'
+  embeddings?: tf.LayerVariable
+
+  constructor(private readonly vocabSize: number,
+    private readonly nEmbd: number, private readonly seed: number) {
+    super({})
+  }
+  override build(): void {
+    this.embeddings = this.addWeight(
+      'wte', //use same name as GPT2
+      [this.vocabSize, this.nEmbd],
+      'float32',
+      tf.initializers.randomNormal({ mean:0, stddev:0.02, seed: this.seed })
+    )
+  }
+
+  override computeOutputShape(inputShape: tf.Shape | tf.Shape[]): tf.Shape | tf.Shape[] {
+    let shape: tf.Shape
+    if (Array.isArray(inputShape) && Array.isArray(inputShape[0])) shape = inputShape[0]
+    else shape = inputShape as tf.Shape
+    // input shape for the token embedding
+    if (shape.length === 2) {
+      // https://github.com/tensorflow/tfjs/blob/3daf152cb794f4da58fce5e21e09e8a4f89c8f80/tfjs-layers/src/layers/embeddings.ts#L155
+      // batch size and sequence length are undetermined 
+      // so the output shape is [null, null, nEmbd]
+      return [...shape, this.nEmbd]
+    }
+    // input shape for the language modeling head
+    // https://github.com/tensorflow/tfjs/blob/3daf152cb794f4da58fce5e21e09e8a4f89c8f80/tfjs-layers/src/layers/core.ts#L258
+    else if (shape.length === 3) {
+      shape[2] = this.vocabSize
+      return shape // [null, null, vocabSize]
+    }
+    else throw new Error('unexpected input shape')
+  }
+
+  override call (input: tf.Tensor | tf.Tensor[], kwargs: Record<string, unknown>): tf.Tensor | tf.Tensor[] {
+    return tf.tidy(() => {
+      if (this.embeddings === undefined) throw new Error('not built')
+      if (Array.isArray(input)) input = input[0]
+      this.invokeCallHook(input, kwargs)
+      
+      // If the input is a 2D tensor, it is a batch of sequences of tokens
+      // so we translate the tokens into embeddings 
+      // using `this.embeddings` as a lookup table
+      if (input.shape.length === 2) {
+        // (batch_size, sequence_length) => (batch_size, sequence_length, nEmbd)
+        return tf.gather(this.embeddings.read(), tf.cast(input, 'int32'), 0)
+      }
+      // If the input is a 3D tensor, it is a sequence of embeddings
+      // so we apply a dense layer to project the embeddings back into the vocabulary space
+      else if (input.shape[2] === this.nEmbd) {
```

Comment:
more in sync with comment #nitpick (also consider using a switch, I find it sometimes more readable)

```suggestion
      else if (input.shape.length === 3) {
```